### ISSUE TITLE
A way to control which request headers are included in the hash for serialized responses in DOM

### ISSUE THREAD

**Issue by Conduitry** [sveltejs/kit#9271]
### Describe the problem

The HTTP headers included in requests made from the server during SSR do not always match those made from the browser. There are multiple reasons why this might happen, including attaching headers with trace/span IDs so that the requests made from the server are connected in a flame graph in observability tools - but you don't want later requests made from the browser to be attached to the same trace.

Since 1.3.6, SvelteKit looks at all of the HTTP headers passed to `fetch` when constructing the hash to compare with the serialized values in the DOM. If any of the headers have changed (even those that I, as a developer, know will not have any bearing on the response), the serialized response will not be used, and the request will be made anew during hydration.

### Describe the proposed solution

I'm not entirely sure. Perhaps we need another callback akin to `filterSerializedResponseHeaders` that lets the developer filter which headers are to be considered when computing the hash. This would probably have to live somewhere where it could be called from both the server and from the browser, because there could well be headers in either place that aren't present in the other.

### Alternatives considered

Some instances of this might be handled by not adding in the extra headers to the arguments of `fetch` but rather injecting them on the fly in the `handleFetch` server hook. This wouldn't help when the extra unmatched headers need to be in the browser.

And it feels like it may not be possible for my case, either. I don't know what headers I want to inject until I'm already about to call `fetch` from my `load` functions, because that's when I (on the server) make the call to create a new child span. I don't see a way to secretly pass that information along to `handleFetch`. I tried mutating the `fetch` instance in `load` by assigning another value on to it, but it seems that the `handleFetch` hook gets a different instance of it. `event.locals` _might_ be a possibility, but only for server `load` functions (at which point I'll know that any fetches inside them are happening from the server anyway), not isomorphic ones.

### Importance

would make my life easier

### Additional Information

I don't currently see a way around this for my app without pinning 1.3.5 for the time being.

**Comment by Conduitry** [comment]
Thinking about this more, injecting the headers during `handleFetch` _might_ still work. I could create the span (and thus generate the span id) from in there and then inject the headers. I'll report back whether I'm successful. It still won't help people who need to have additional headers on the client side only. But how antsy I personally am for this feature will definitely depend on whether I am stuck on 1.3.5.

**Comment by Conduitry** [comment]
Yeah, I think I might be good. What I'm testing right now is (roughly)

```js
export const handleFetch = ({ request, fetch }) => {
	return create_span((span) => {
		const headers = Object.fromEntries(request.headers);
		inject_headers_from_span_into(span, headers);
		return fetch(new Request(request, { headers }));
	});
};
```

(with service-specific calls replaced with generic-sounding functions). This appears to be working. In my shared API client thing, I can then stop creating the span and can just call `fetch` with the same headers whether I'm on the server or the browser, and the hashing/caching works as desired.

**Comment by SilverMira** [comment]
Facing this issue, but instead my scenario can only be resolved by either allowing headers set from within `handleFetch` to be considered when hashing, or allow the client side's hashing to ignore certain headers.

I'm setting `Authorization` header from a cookie passed by a client in `handleFetch` when making API request upstream, but this header is not considered when hashing. When the client hydrates, the client side fetches the same resource with the same `Authorization` value, but this time it's considered when hashing, causing a mismatch.

**Comment by maxime-carbonneau** [comment]
> Facing this issue, but instead my scenario can only be resolved by either allowing headers set from within `handleFetch` to be considered when hashing, or allow the client side's hashing to ignore certain headers.
> 
> I'm setting `Authorization` header from a cookie passed by a client in `handleFetch` when making API request upstream, but this header is not considered when hashing. When the client hydrates, the client side fetches the same resource with the same `Authorization` value, but this time it's considered when hashing, causing a mismatch.

I face the same issue. I « solved » it by creating a store and changing it value in onMount.

```javascript
// lib/stores.js
export const hydrationStore = writable(true);
```
```javascript
// +layout.svelte
import { onMount } from 'svelte';
import { hydrationStore } from "$lib/stores";

onMount(() => {
    $hydrationStore = false;
});

```
```javascript
// +page.js
import { hydrationStore } from "$lib/stores";
import { browser } from '$app/environment';

let hydration;
if (browser) {
    hydration.subscribe(value => hydration = value);
}

export async function load(event) {
    if (browser && hydration) {
        // Do the fetch without Authorization, so it will reuse data from hashing
        fetch(...);
    }
    ...
}
```

**Comment by coryvirok** [comment]
> Facing this issue, but instead my scenario can only be resolved by either allowing headers set from within `handleFetch` to be considered when hashing, or allow the client side's hashing to ignore certain headers.
> 
> I'm setting `Authorization` header from a cookie passed by a client in `handleFetch` when making API request upstream, but this header is not considered when hashing. When the client hydrates, the client side fetches the same resource with the same `Authorization` value, but this time it's considered when hashing, causing a mismatch.

I'm facing this exact same issue and use-case. I have an external TRPC server that uses the Authorization header. In my Kit instance, I have a TRPC client in the browser that uses a JWT from the server to make authenticated TRPC calls. However, when I make authenticated TRPC calls on the server (by injecting the Authorization header in handleFetch) the Authorization header never makes its way into the hash computed for the server fetch.

This means there's the exact same TRPC call done on the server and on the client during hydration. 

I've been scratching my head for a few hours trying to find a workaround. Has anyone figured out a way?

**Comment by Rich-Harris** [comment]
Closing per https://github.com/sveltejs/kit/pull/11008#issuecomment-3197761150
